velox/functions/prestosql/tests/ArrayNGramTest.cpp

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <optional>
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+using namespace facebook::velox::test;
+using namespace facebook::velox::functions::test;
+
+namespace {
+
+class ArrayNGramTest : public FunctionBaseTest {
+ protected:
+  template <typename T>
+  void testNgram(
+      const std::vector<std::vector<std::optional<T>>>& inputArray,
+      int64_t size,
+      const std::vector<std::optional<std::vector<std::optional<T>>>>&
+          expectedOutput) {
+    auto input = makeNullableArrayVector<T>({inputArray});
+    auto result =
+        evaluate(fmt::format("ngrams(c0, {})", size), makeRowVector({input}));
+
+    auto expected = makeNullableNestedArrayVector<T>({{expectedOutput}});
+    assertEqualVectors(expected, result);
+  }
+};
+
+TEST_F(ArrayNGramTest, intArrays) {
+  testNgram<int64_t>({{1, 2, 3, 4}}, 1, {{{1}}, {{2}}, {{3}}, {{4}}});
+  testNgram<int64_t>({{1, 2, 3, 4}}, 2, {{{1, 2}}, {{2, 3}}, {{3, 4}}});
+  testNgram<int64_t>({{1, 2, 3, 4}}, 3, {{{1, 2, 3}}, {{2, 3, 4}}});
+  testNgram<int64_t>({{1, 2, 3, 4}}, 4, {{{1, 2, 3, 4}}});
+  testNgram<int64_t>({{1, 2, 3, 4}}, 5, {{{1, 2, 3, 4}}});
+  testNgram<int64_t>(
+      {{1, 2, 3, 4}}, std::numeric_limits<int32_t>::max(), {{{1, 2, 3, 4}}});
+  testNgram<int64_t>(
+      {{1, 2, 3, 4}},
+      std::numeric_limits<int32_t>::max() + (int64_t)(1000),
+      {{{1, 2, 3, 4}}});
+  testNgram<int64_t>({{}}, 1, {{{}}});
+  testNgram<int64_t>({{}}, 10, {{{}}});
+
+  // Test of invalid N
+  auto input = makeArrayVector<int64_t>({{1, 2, 3, 4}});
+  VELOX_ASSERT_THROW(
+      evaluate("ngrams(c0, 0)", makeRowVector({input})), "N must be positive.");
+  VELOX_ASSERT_THROW(
+      evaluate("ngrams(c0, -5)", makeRowVector({input})),
+      "N must be positive.");
+  input = makeArrayVector<int64_t>({{}});
+  VELOX_ASSERT_THROW(
+      evaluate("ngrams(c0, 0)", makeRowVector({input})), "N must be positive.");
+}
+
+TEST_F(ArrayNGramTest, stringArrays) {
+  testNgram<std::string>(
+      {{"foo", "bar", "baz", "foo"}},
+      1,
+      {{{"foo"}}, {{"bar"}}, {{"baz"}}, {{"foo"}}});
+  testNgram<std::string>(
```

Comment:
Please add tests where the arrays have a mix of 
i) Strings with > 12 characters in size for non-inline coverage.
ii) Strings have special characters as well.